use discovery selectors to filter namespaces

jmazzitelli opened the pull request:
fixes: https://github.com/kiali/kiali/issues/3914

operator PR: https://github.com/kiali/kiali-operator/pull/628

The Istio ConfigMap has discoverySelectors that Kiali will now use to filter namespaces.

If no discoverySelectors are specified, or discoverySelectors is specified as an empty list in the ConfigMap, then Kiali works as before.

If discoverySelectors are specified, then they will filter out any namespaces that do not match. After that filter is performed, Kiali will further exclude namespaces defined in the excludes config settings in the Kiali CR (this works as before).

nrfox commented on business/namespaces.go:
```diff
@@ -78,6 +80,26 @@ func (in *NamespaceService) GetNamespaces(ctx context.Context) ([]models.Namespa
+		istioConfig = ic
+	}
+
+	if istioConfig.DiscoverySelectors != nil && len(istioConfig.DiscoverySelectors) > 0 {
```
Just trying to get your tests passing... you likely need an additional check for `istioConfig != nil`

jmazzitelli replied:
Well, this is what I'm struggling with. In production, the Kiali server should always have access to the istio configmap - that value should never be nil (or, technically, you should never get here if its nil). This function aborts with an error if the configmap cannot be processed - so you should never get here.

Obviously, I could just not abort and keep going with a nil ConfigMap but:

(a) that would solely be to support unit testing and is not valid in production - and that sucks to have production code written in a way solely to support unit tests, especially when it really isn't "right"

(b) IMO, if we can't get the DS, then we need to abort. This could be considered a security issue. Because suppose Kiali is configured with ** but the user wants to only make available those namespaces filtered by DS - if we do not have access to what the DSs are - we should NOT default to all namespace (i.e. we should not default to a list of namespaces that were not filtered by DS). To me, that is a security flaw. If we can't get to the DS, then we need to abort and tell the user we are broken until Kiali has access to see the Istio ConfigMap so it can correctly display only those namespaces we should be.

nrfox replied:
> Well, this is what I'm struggling with. In production, the Kiali server should always have access to the istio configmap - that value should never be nil (or, technically, you should never get here if its nil). This function aborts with an error if the configmap cannot be processed - so you should never get here.

Kiali may never have access to external control planes or istio may temporarily be down in which case Kiali should still keep functioning.

> This could be considered a security issue.

I thought DS was not about access? Isn't that what `cluster_wide_access` is for? DS is not a security boundary for istio either I don't think.

jmazzitelli replied:
If we all don't think this is a security-issue to worry about, I can relax that code. I'm security-paranoid because I know all the times people have complained to us that we expose more than we should, which is why I was hesitant.

jmazzitelli replied:
This has been reworked so the server continues on even if the Istio ConfigMap is not found or for some reason can't be parsed. It will still log an error but it keeps going.

jmazzitelli replied:
lo-and-behold almost all the unit tests pass again :)

There is one failure that I don't understand:
```
--- FAIL: TestTlsPerfNsDr (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x18022ae]

goroutine 8785 [running]:
testing.tRunner.func1.2({0x1c0ef60, 0x311c200})
	/home/jmazzite/bin/go-install/go/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	/home/jmazzite/bin/go-install/go/src/testing/testing.go:1399 +0x39f
panic({0x1c0ef60, 0x311c200})
	/home/jmazzite/bin/go-install/go/src/runtime/panic.go:884 +0x212
github.com/kiali/kiali/kubernetes/cache.(*kialiCacheImpl).GetConfigMap(0xc000818790?, {0x1ef387a?, 0x1ef1e16?}, {0x1ee9965?, 0xc000b42a00?})
	<autogenerated>:1 +0x2e
github.com/kiali/kiali/business.(*NamespaceService).GetNamespaces(0xc000a30e70, {0x21fa8f0, 0xc0001ac008})
	/home/jmazzite/source/kiali/kiali/business/namespaces.go:86 +0x278
github.com/kiali/kiali/business.(*TLSService).getNamespaces(0x21fa8f0?, {0x21fa8f0?, 0xc0001ac008?})
```


I don't get it because this is wrapped in the kialiCache != nil test, so something else must be blowing up

jmazzitelli replied:
this test has been fixed with a not-so-obvious change.

jmazzitelli commented:
@nrfox I need suggestions to know how to use the new "cluster wide access" setting - the cache will need to use it. (in fact I forgot to add that to the config.go in this PR - I'll do that next). But I need to know where/how the cache will make use of it.

nrfox commented:
> @nrfox I need suggestions to know how to use the new "cluster wide access" setting - the cache will need to use it. (in fact I forgot to add that to the config.go in this PR - I'll do that next). But I need to know where/how the cache will make use of it.

I believe you just need to change this method to return the new setting instead of looking for `**` in AN: https://github.com/kiali/kiali/blob/3b4ecc9c5153bd7c12f7178146f3d8592db3ae4b/config/config.go#L786-L796

jmazzitelli commented:
OK, latest commit updates that function and sets the defaults in a way that unit tests continue to pass.

jmazzitelli commented:
all operator molecule tests pass with this PR applied to the server image

jmazzitelli commented:
One note that needs to be mentioned. The main difference the user will notice between the implementation where the operator processes discoverySelectors, aka DS (the previous WIP implementation) and now where the server processes DS, the user will have to know to create an AN (accessible_namespaces list) that at minimum includes the namespaces selected by the DS. Is this a burden on the user?

jshaughn commented:
> The main difference the user will notice between the implementation where the operator processes discoverySelectors, aka DS (the previous WIP implementation) and now where the server processes DS, the user will have to know to create an AN (accessible_namespaces list) that at minimum includes the namespaces selected by the DS. Is this a burden on the user?

I think the new approach may just require the same from users as we did in the past.  Existing users already using an explicit AN should be fine or using ** should also be fine, right?  So this is mainly for new users, I think.  And I think the guidance is the same, If using namespace-security then you need the specific AN list.  If cluster-wide you can have an explicit AN or not, up to you.  Either way, the DS definitions, if any, can only narrow the accessible namepaces.

nrfox reviewed (commented):
One question otherwise LGTM

jmazzitelli commented on business/namespaces.go:
```diff
@@ -78,6 +80,26 @@ func (in *NamespaceService) GetNamespaces(ctx context.Context) ([]models.Namespa
+	// determine what the discoverySelectors are by examining the Istio ConfigMap
+	var discoverySelectors []*meta_v1.LabelSelector
+	if kialiCache != nil {
+		if icm, err := kialiCache.GetConfigMap(configObject.IstioNamespace, configObject.ExternalServices.Istio.ConfigMapName); err == nil {
```
Good point.. do we have a function that can be used to isolate the configmap name? Something like "getIstioConfigMap" and have it deal with the versioning/canary details? I believe with canary deployments, it also uses a differently named config map (??)

@leandroberetta might know, he did some canary work in the past ^^

nrfox replied:
> Good point.. do we have a function that can be used to isolate the configmap name? Something like "getIstioConfigMap" and have it deal with the versioning/canary details?

That would be useful. Probably can be added to the config object.

leandroberetta replied:
It's a good point, every time you install a canary, you'll get a new config map for that revision. Particulary in the canary work I've done, I'm not using the config map, but there are places that the config map is used. 

So, after a canary, eventually, Kiali will need to grab the new configmap (has a different name). I will create an issue to document this better, and to see if it is just that.

nrfox reviewed (approved):
LGTM. I think this needs an e2e test but I'll add the label and it can be included with the PR or after.

Xunzhuo reviewed (approved):
Overall looks good to me, thanks

jshaughn reviewed (approved):
This looks OK to me.  The '**' use is an internal impl detail from what I can see.  We want to remove '**' from any public exposure as it's very confusing and should be replaced by `cluster_wide_access: true`.

jshaughn commented:
Added DNM label because the Operator PR may need some work.

jshaughn commented:
@Xunzhuo Just a question, do you use istio Discovery Selectors in your environment?

Xunzhuo commented:
@jshaughn yes, we use it for multi Istio control plane in a single cluster

jmazzitelli commented:
fixed a conflict that was caused by a major refactoring if istio_test.go. I squashed the commits in this PR so conflicts won't be difficult next time. But I confirmed the code is the same as it was before the squash.

jshaughn commented on config/config.go:
```diff
@@ -361,6 +361,7 @@ type OpenIdConfig struct {
 // DeploymentConfig provides details on how Kiali was deployed.
 type DeploymentConfig struct {
 	AccessibleNamespaces []string `yaml:"accessible_namespaces"`
+	ClusterWideAccess    bool     `yaml:"cluster_wide_access,omitempty"`
```
I believe with `omitempty` a bool set to `false` will be omitted.   As long as unset would always reflect `false` I guess this is OK, but given the semantics of CWA I'm wondering if it's best to not omit.

jmazzitelli replied:
`omitempty` is meaningless in our context and not used. In fact, we could remove every mention of `omitempty` in this file and it would do nothing to us because it is only used for marshaling into YAML and AFAIK we do not marshal into yaml (we only unmarshal from the ConfigMap to the internal Go structs - we don't go the other way around and go from the structs to a YAML string - we certainly don't write back to the ConfigMap nor write to a yaml file anywhere).

See: https://pkg.go.dev/gopkg.in/yaml.v2#Marshal

```
omitempty    Only include the field if it's not set to the zero
             value for the type or to empty slices or maps.
             Zero valued structs will be omitted if all their public
             fields are zero, unless they implement an IsZero
             method (see the IsZeroer interface type), in which
             case the field will be excluded if IsZero returns true.
```

jshaughn replied:
It's fine if we're never going to marshal to YAML.  But in general, for booleans, I've been burned when using omitempty, so it was just a mention.

jshaughn reviewed (approved):
Code looks fine.  DS are narrowing AN as expected.  LG2M.
